pulp_rpm/app/tasks/synchronizing.py

```diff
@@ -1583,31 +1591,40 @@ async def run(self):
                     content_q_by_type[model_type] = content_q_by_type[model_type] | unit_q
                     d_content_by_nat_key[d_content.content.natural_key()].append(d_content)
 
-            for model_type, content_q in content_q_by_type.items():
-                try:
-                    await sync_to_async(model_type.objects.filter(content_q).touch)()
-                except AttributeError:
-                    raise TypeError(
-                        "Plugins which declare custom ORM managers on their content classes "
-                        "should have those managers inherit from "
-                        "pulpcore.plugin.models.ContentManager."
-                    )
+            # For each entry in the batch, determine if we need to "fix up" the data before
+            # letting it pass down the pipeline.
+            #
+            # NOTE that we "remember" entities we've touched, and **which fields** we've touched,
+            # in order to persist **just the changes we make** below. This helps us avoid some
+            # nasty deadlock-windows between save() and touch() in this stage.
+            #
+            # IF/WHEN/AS YOU ADD TO THIS CODE - make sure you update modified_results_by_pk and
+            # modified_fields_by_pk, or your changes WILL NOT BE PERSISTED.
+
+            # holds the to-be-saved results by-pk
+            modified_results_by_pk = {}
```

Comment:
Can we move all of this and the saving bits inside the loop, so that it's reset when dealing with different model types